lib/compress/zstd_compress.c

```diff
@@ -3950,9 +3950,14 @@ size_t ZSTD_compressStream2( ZSTD_CCtx* cctx,
 
 
 #ifdef ZSTD_MULTITHREAD
+
+/* Invoke fuzzing every time for mt even on small files */
```

Comment:
I may have missed something here.

This code is defined to not use the ZSTDMT path when input size is too small.
What do we gain by forcing the fuzzer to test a path that is never used in production?

Reply:
The original rationale was to get it to catch the refPrefix bug from https://github.com/facebook/zstd/pull/2088 but it didn't do that

>What do we gain by forcing the fuzzer to test a path that is never used in production?

Well, it exposed the heap-overflow above. I suppose that code path is one that should already be covered but this new ifndef caught it immediately where as without it, we hadn't caught it so far. The rationale here is that by forcing the fuzzer into more paths than are used in production, we might give it more opportunities to find bugs that are in production code paths.

Reply:
This is not about using a code path which is "rarely" used in production, which would be nonetheless worth fuzzing for extended coverage, just in case one scenario might actually trigger it.

This is about a code path which is _expressly forbidden_, and therefore _impossible_ to reach.

There are conditions for this function to work. They are safely checked. Why being surprised that it doesn't work anymore once the protections are removed ?

To give another example, one doesn't just test `memcpy()` with overlapping buffers and conclude that it's buggy: tests still have to respect the working conditions of the tested functions,  and `memcpy()` interface expressly forbids such a use case.
This is the same here.

Reply:
@Cyan4973 I think it is worth fixing the conditions to ensure we can run in multithreaded mode with smaller inputs. Larger inputs are slower to fuzz, and are not provided as often by oss-fuzz. Making it easier for the fuzzer to get into this code path should improve our coverage of zstdmt, and hopefully find real bugs.